### PR TITLE
Handling Volume at early stage

### LINKED ISSUE #544: Fatal error saying "/etc/nginx/certs": must be unique
I'm using `jrcs/letsencrypt-nginx-proxy-companion` in my `docker-compose.production.yml` and `kompose -f docker-compose.production.yml up` throws an error saying

```
WARN Unsupported logging key - ignoring
WARN Volume mount on the host "/var/run/docker.sock" isn't supported - ignoring path on the host
WARN Volume mount on the host "./certs" isn't supported - ignoring path on the host
WARN Volume mount on the host "/var/run/docker.sock" isn't supported - ignoring path on the host
WARN Volume mount on the host "./certs" isn't supported - ignoring path on the host
WARN Volume mount on the host "/var/run/docker.sock" isn't supported - ignoring path on the host
WARN Volume mount on the host "./certs" isn't supported - ignoring path on the host
INFO We are going to create Kubernetes Deployments, Services and PersistentVolumeClaims for your Dockerized application. If you need different kind of resources, use the 'kompose convert' and 'kubectl create -f' commands instead.

INFO Successfully created Service: letsencrypt
INFO Successfully created Service: proxy
FATA Error while deploying application: Deployment.extensions "letsencrypt" is invalid: spec.template.spec.containers[0].volumeMounts[4].mountPath: Invalid value: "/etc/nginx/certs": must be unique
```

How can I fix this issue?

---

My `kompose version` is `0.5.0 ()`

And my `docker-compose.production.yml` below.

```yml
version: "2"

services:

  proxy:
    image: jwilder/nginx-proxy
    privileged: true
    container_name: proxy
    ports:
      - 80:80
      - 443:443
    volumes:
      - /etc/nginx/vhost.d
      - /var/run/docker.sock:/tmp/docker.sock:ro
      - ./certs:/etc/nginx/certs:ro
      - /usr/share/nginx/html
    restart: always
    logging:
      options:
        max-size: 5m
        max-file: "10"

  letsencrypt:
    image: jrcs/letsencrypt-nginx-proxy-companion
    privileged: true
    container_name: ssl
    volumes_from:
      - proxy
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - ./certs:/etc/nginx/certs:rw

networks:
  default:
    external:
      name: shared
```

### PR DESCRIPTION
**Pull request by surajnarwade**
It will resolve #544 as well as refactor volume handling part.

### CODE REVIEW

**Comment by surajnarwade** [comment]
this PR is still in WIP, so I will add tests later.
@cdrage @kadel @surajssd suggestions needed

**Comment by cdrage** [comment]
@surajnarwade DaemonSet's aren't working. See: https://travis-ci.org/kubernetes-incubator/kompose/jobs/237507721#L398

Please get current tests to pass as well :smile: before we add new tests!

**Comment by surajssd** [comment]
@surajnarwade please rebase

**Comment by surajnarwade** [comment]
@surajssd @cdrage  done with rebase and tests are green now, needs your review please :)

File: pkg/kobject/kobject.go [hunk 1]
```diff
@@ -107,3 +108,15 @@ type Ports struct {
...
+type Volumes struct {
+	SvcName    string
+	MountPath  string
+	VFrom      string
+	VolumeName string
+	Host       string
+	Container  string
+	Mode       string
+	PVCName    string
+}
```
Comment (cdrage) [inline pkg/kobject/kobject.go]: I don't like how many variables there are. It may be better copying how they do it within Docker/CLI so it's easier for us to be compatible with them. See: https://github.com/docker/cli/blob/master/cli/compose/types/types.go#L243 or perhaps even import and use the type.

File: pkg/transformer/kubernetes/kubernetes.go [hunk 2]
```diff
@@ -376,56 +377,51 @@ func (k *Kubernetes) ConfigVolumes(name string, service kobject.ServiceConfig) (
...
-	for _, volume := range service.Volumes {
-
-		volumeName, host, container, mode, err := transformer.ParseVolume(volume)
-		if err != nil {
-			log.Warningf("Failed to configure container volume: %v", err)
-			continue
-		}
-
-		log.Debug("Volume name %s", volumeName)
+	//POC work ############################
```
Comment (cdrage) [inline pkg/transformer/kubernetes/kubernetes.go]: remove this comment

File: pkg/loader/compose/v1v2.go [hunk 3]
```diff
@@ -268,6 +269,50 @@ func libComposeToKomposeMapping(composeObject *project.Project) (kobject.Kompose
...
+			if vol.VFrom != "" {
+				for _, j := range komposeObject.ServiceConfigs[vol.VFrom].Vols {
+					if vol.MountPath == j.MountPath {
+						vol.PVCName = j.PVCName
+					}
+				}
+			}
+		}
+	}
+
```
Comment (cdrage) [inline pkg/loader/compose/v1v2.go]: Is it possible to refactor this into it's own function (parseV1V2Volumes maybe?)

File: pkg/transformer/kubernetes/kubernetes.go [hunk 4]
```diff
@@ -376,56 +377,51 @@ func (k *Kubernetes) ConfigVolumes(name string, service kobject.ServiceConfig) (
...
+	for _, vol1 := range service.Vols {
```
Comment (cdrage) [inline pkg/transformer/kubernetes/kubernetes.go]: please change vol1 variable name to something better, why not `volume` like before?

File: pkg/kobject/kobject.go [hunk 5]
```diff
@@ -92,6 +92,7 @@ type ServiceConfig struct {
 	MemLimit        yaml.MemStringorInt `compose:"mem_limit" bundle:""`
 	TmpFs           []string            `compose:"tmpfs" bundle:""`
 	Dockerfile      string              `compose:"dockerfile" bundle:""`
+	Vols            []Volumes           `compose:"" bundle:""`
```
Comment (surajnarwade) [inline pkg/kobject/kobject.go]: `Volumes` is already there  https://github.com/kubernetes-incubator/kompose/blob/master/pkg/kobject/kobject.go#L70 for libcompose mapping, `Vols` is new struct for handling volumes.

**Comment by surajnarwade** [comment]
@cdrage , in order to answer your question regarding copying functionality from `docker/cli`, it can be used moreover for root level `volumes` key but not the thing that I am working on.

**Comment by cdrage** [comment]
@surajnarwade I believe that you can use similar type.

See how we "copy" from libcompose in kobject.go in terms of ServiceConfig. This should be the same in terms of Volumes. It makes sense to try to keep the underlying code similar to upstream code.

What do you think @kadel ? Also regarding the many variables (see https://github.com/kubernetes-incubator/kompose/pull/626#discussion-diff-122452810R120)

File: pkg/kobject/kobject.go (hunk 5 above)
Comment (surajssd) [inline pkg/kobject/kobject.go]: I think it is not a good idea to have both `Volumes` and `Vols`, please remove the `[]String`

File: pkg/kobject/kobject.go [hunk 6]
```diff
@@ -107,3 +108,15 @@ type Ports struct {
...
 }
+
+// Volumes holds the volume struct of container
+type Volumes struct {
+	SvcName    string
+	MountPath  string
+	VFrom      string
+	VolumeName string
+	Host       string
+	Container  string
```
Comment (surajssd) [inline pkg/kobject/kobject.go]: Is this container mount Path ?

File: pkg/kobject/kobject.go (hunk 1 above)
Comment (surajssd) [inline pkg/kobject/kobject.go]: @cdrage I think this is needed for doing the whole volumesFrom thing reliably, since these will be resolved once to save computation later.

File: pkg/loader/compose/v1v2.go [hunk 7]
```diff
@@ -277,3 +322,80 @@ func checkLabelsPorts(noOfPort int, labels string, svcName string) error {
...
+		if vols.MountPath == v.MountPath {
+			return true
+		}
+	}
+	return false
+}
+
+func handleVolume(svcName string, komposeObject kobject.KomposeObject) (volume []kobject.Volumes) {
+
+	if komposeObject.ServiceConfigs[svcName].VolumesFrom != nil {
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: please put in lots of comments since it is hard to understand what is happening here

Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: After certain time even the programmer forgets why s/he wrote a piece of code! :-D

Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: comments not just here but in other parts of code

File: pkg/loader/compose/v1v2.go [hunk 8]
```diff
@@ -277,3 +322,80 @@ func checkLabelsPorts(noOfPort int, labels string, svcName string) error {
...
+		for _, depVol := range komposeObject.ServiceConfigs[svcName].VolumesFrom {
+
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: nit: `s///n`

File: pkg/loader/compose/v1v2.go [hunk 9]
```diff
@@ -277,3 +322,80 @@ func checkLabelsPorts(noOfPort int, labels string, svcName string) error {
...
+			dVols := handleVolume(depVol, komposeObject)
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: also why this recursive call is made, needs an explanation

File: pkg/kobject/kobject.go (hunk 5 above)
Comment (surajssd) [inline pkg/kobject/kobject.go]: What I mean here is that `[]Volumes` was added to get rid of `[]string` ?

File: pkg/loader/compose/v1v2.go [hunk 10]
```diff
@@ -277,3 +322,80 @@ func checkLabelsPorts(noOfPort int, labels string, svcName string) error {
...
+			var cVols []kobject.Volumes
+			for _, v := range komposeObject.ServiceConfigs[svcName].Volumes {
+				vol := kobject.Volumes{
+					SvcName:   svcName,
+					MountPath: v,
+				}
+				cVols = append(cVols, vol)
+			}
+			for _, cv := range cVols {
+				i, dv := getVol(cv, dVols)
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: nit for general practices reason, `s/i/ok` ?

File: pkg/loader/compose/v1v2.go [hunk 11]
```diff
@@ -277,3 +322,80 @@ func checkLabelsPorts(noOfPort int, labels string, svcName string) error {
...
+				if i == true {
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: also in a conditional where you are using boolean values no need to check it with booleans! essentially I mean `s/i == true/i`

File: pkg/loader/compose/v1v2.go [hunk 12]
```diff
@@ -277,3 +322,80 @@ func checkLabelsPorts(noOfPort int, labels string, svcName string) error {
...
+				}
+				volume = append(volume, cv)
+			}
+
+			for _, dv := range dVols {
+				if checkvolpresent(dv, cVols) {
+					dv.VFrom = dv.SvcName
+					volume = append(volume, dv)
+				}
+
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: nit: `s///n`

File: pkg/loader/compose/v1v2.go [hunk 13]
```diff
@@ -277,3 +322,80 @@ func checkLabelsPorts(noOfPort int, labels string, svcName string) error {
...
+			}
+
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: nit: `s///n`

File: pkg/loader/compose/v1v2.go [hunk 14]
```diff
@@ -277,3 +322,80 @@ func checkLabelsPorts(noOfPort int, labels string, svcName string) error {
...
+		}
+	} else {
+		for _, v := range komposeObject.ServiceConfigs[svcName].Volumes {
+			vol := kobject.Volumes{
+				SvcName:   svcName,
+				MountPath: v,
+			}
+			volume = append(volume, vol)
+
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: nit: `s///n`

File: pkg/loader/compose/v1v2.go [hunk 15]
```diff
@@ -277,3 +322,80 @@ func checkLabelsPorts(noOfPort int, labels string, svcName string) error {
...
+func checkvolpresent(tofind kobject.Volumes, cVols []kobject.Volumes) bool {
+	for _, cv := range cVols {
+		if tofind.MountPath != cv.MountPath {
+			return true
+		}
+	}
+	return false
+}
+
+func getVol(tofind kobject.Volumes, dVols []kobject.Volumes) (bool, kobject.Volumes) {
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: This is a generic function that searches array of volumes without any bias on what the array of volumes is so maybe name it better, like `s/dVols/vols/`

File: pkg/kobject/kobject.go (hunk 6 above)
Comment (surajssd) [inline pkg/kobject/kobject.go]: I mean I am confused on `MountPath` vs `Container` ?

Comment (surajnarwade) [inline pkg/kobject/kobject.go]: yeah

File: pkg/loader/compose/v1v2.go [hunk 16]
```diff
@@ -277,3 +322,80 @@ func checkLabelsPorts(noOfPort int, labels string, svcName string) error {
...
+			}
+			volume = append(volume, vol)
+
+		}
+	}
+	return volume
+}
+
+func checkvolpresent(tofind kobject.Volumes, cVols []kobject.Volumes) bool {
+	for _, cv := range cVols {
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: I don't see a reason to have this function? Because `getVol` does similar thing but with inverted functionality can we use just one?

Comment (surajnarwade) [inline pkg/loader/compose/v1v2.go]: but `getVol` returns with extra `Volumes` struct

File: pkg/kobject/kobject.go (hunk 6 above)
Comment (surajnarwade) [inline pkg/kobject/kobject.go]: `mountpath` can be `host:container`, hence `container` is that directory, for example, `/abc`

File: pkg/kobject/kobject.go (hunk 5 above)
Comment (surajnarwade) [inline pkg/kobject/kobject.go]: either I can swap names but we can't remove []string

**Comment by surajnarwade** [comment]
@surajssd review please :)

File: pkg/kobject/kobject.go [hunk 17]
```diff
@@ -92,6 +92,7 @@ type ServiceConfig struct {
 	MemLimit        yaml.MemStringorInt `compose:"mem_limit" bundle:""`
 	TmpFs           []string            `compose:"tmpfs" bundle:""`
 	Dockerfile      string              `compose:"dockerfile" bundle:""`
+	Volumes         []Volumes           `compose:"" bundle:""`
```
Comment (cdrage) [inline pkg/kobject/kobject.go]: Volumes variable isn't used anymore?

File: pkg/loader/compose/v1v2.go [hunk 18]
```diff
@@ -268,6 +269,51 @@ func libComposeToKomposeMapping(composeObject *project.Project) (kobject.Kompose
...
+			if vol.VFrom == "" {
+				vol.PVCName = fmt.Sprintf("%s-claim%d", vol.SvcName, count)
+				count++
+			}
+
+		}
+		var temp = komposeObject.ServiceConfigs[name]
+		temp.Volumes = FinalVol
+		komposeObject.ServiceConfigs[name] = temp
+	}
```
Comment (cdrage) [inline pkg/loader/compose/v1v2.go]: There's no way we can add this much code into this fuction, could you refactor this into a separate function (see my other comment too)

File: pkg/loader/compose/v1v2.go [hunk 19]
```diff
@@ -277,3 +323,88 @@ func checkLabelsPorts(noOfPort int, labels string, svcName string) error {
...
+//returns all volumes associated with service, if `volumes_from` key is used, we have to retrieve volumes from the services which are mentioned there. Hence, recursive function is used here.
+func handleVolume(svcName string, komposeObject kobject.KomposeObject) (volume []kobject.Volumes) {
+	//if volumes-from key is present
+	if komposeObject.ServiceConfigs[svcName].VolumesFrom != nil {
+		//iterating over services from `volumes-from`
+		for _, depSvc := range komposeObject.ServiceConfigs[svcName].VolumesFrom {
+			//recursive call for retrieving volumes of services from `volumes-from`
+			dVols := handleVolume(depSvc, komposeObject)
+			var cVols []kobject.Volumes
+			//creating array of volumes of current service(svcName)
```
Comment (cdrage) [inline pkg/loader/compose/v1v2.go]: A lot of your comments don't have the spacing after the `//`. 

Please add some new-lines too so that the separate sections are spaced out a bit better.

File: pkg/kobject/kobject.go (hunk 17 above)
Comment (cdrage) [inline pkg/kobject/kobject.go]: Nvm, ignore my comment. Just the functionality has changed.

File: pkg/loader/compose/v1v2.go [hunk 20]
```diff
@@ -268,12 +269,146 @@ func libComposeToKomposeMapping(composeObject *project.Project) (kobject.Kompose
...
+			}
+			log.Debug("Volume name %s", vol.VolumeName)
+
+			if vol.VFrom == "" {
+				vol.PVCName = fmt.Sprintf("%s-claim%d", vol.SvcName, count)
+				count++
+			}
+
+		}
+		var temp = komposeObject.ServiceConfigs[name]
```
Comment (kadel) [inline pkg/loader/compose/v1v2.go]: this whole block needs comments describing what is happening here

File: pkg/kobject/kobject.go [hunk 21]
```diff
@@ -67,7 +67,7 @@ type ServiceConfig struct {
 	Command         []string            `compose:"command" bundle:"Command"`
 	WorkingDir      string              `compose:"" bundle:"WorkingDir"`
 	Args            []string            `compose:"args" bundle:"Args"`
-	Volumes         []string            `compose:"volumes" bundle:"Volumes"`
+	VolList         []string            `compose:"volumes" bundle:"Volumes"`
```
Comment (kadel) [inline pkg/kobject/kobject.go]: what is a difference between `VolList` and `Volumes`? What is purpose of `VolList`? We need comments describing both

File: pkg/kobject/kobject.go [hunk 22]
```diff
@@ -107,3 +108,15 @@ type Ports struct {
 	HostIP        string
 	Protocol      api.Protocol
 }
+
+// Volumes holds the volume struct of container
+type Volumes struct {
+	SvcName    string
+	MountPath  string
+	VFrom      string
```
Comment (kadel) [inline pkg/kobject/kobject.go]: What is `VFrom` ?

File: pkg/kobject/kobject.go (hunk 1 above)
Comment (kadel) [inline pkg/kobject/kobject.go]: This is a bit confusing for me also. Maybe it would be nice to have some comments describing what some of the variables are for.

File: script/test/fixtures/v3/output-os-full-example.json [hunk 23]
```diff
@@ -423,138 +360,6 @@
       "status": {
         "dockerImageRepository": ""
       }
-    },
-    {
-      "kind": "PersistentVolumeClaim",
```
Comment (kadel) [inline script/test/fixtures/v3/output-os-full-example.json]: Why are those PVC removed from test?

File: script/test/fixtures/service-name-change/output-os.json [hunk 24]
```diff
@@ -246,21 +245,21 @@
...
-                  "claimName": "servicenamechange-wordpress-data"
+                  "claimName": ""
                 }
               },
               {
-                "name": "servicenamechange-apache-data",
+                "name": "",
                 "persistentVolumeClaim": {
-                  "claimName": "servicenamechange-apache-data"
+                  "claimName": ""
```
Comment (kadel) [inline script/test/fixtures/service-name-change/output-os.json]: empty name?

File: pkg/kobject/kobject.go (hunk 21 above)
Comment (surajnarwade) [inline pkg/kobject/kobject.go]: added description in comment section

**Comment by surajnarwade** [comment]
@kadel @cdrage , I have updated the PR with comments as suggested, needs review :)

File: script/test/fixtures/multiple-compose-files/output-k8s.json [hunk 25]
```diff
@@ -236,28 +226,6 @@
         }
       },
       "status": {}
-    },
-    {
-      "kind": "PersistentVolumeClaim",
```
Comment (kadel) [inline script/test/fixtures/multiple-compose-files/output-k8s.json]: Why is PVC removed?

Comment (surajnarwade) [inline script/test/fixtures/multiple-compose-files/output-k8s.json]: @kadel , it;s not removed, just order misplaced :)

File: pkg/loader/compose/v1v2.go [hunk 26]
```diff
@@ -268,12 +269,154 @@ func libComposeToKomposeMapping(composeObject *project.Project) (kobject.Kompose
...
+		for _, v := range vols {
+			if !findDuplicateVol(v, FinalVol) {
+				FinalVol = append(FinalVol, v)
+			}
+		}
+		var count int
+
+		// Parse volume information and store it in Volumes struct
+		for v, _ := range FinalVol {
+			vol := &FinalVol[v]
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: why is this done?

File: pkg/loader/compose/v1v2.go [hunk 27]
```diff
@@ -268,12 +269,154 @@ func libComposeToKomposeMapping(composeObject *project.Project) (kobject.Kompose
...
+				}
+			}
+		}
+	} else {
+		// if `volumes-from` is not present
+		for _, v := range komposeObject.ServiceConfigs[svcName].VolList {
+			vol := kobject.Volumes{
+				SvcName:   svcName,
+				MountPath: v,
+			}
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: You can use `transformer.ParseVolume` here, then you won't need to call it later in the final volume.

File: pkg/loader/compose/v1v2.go [hunk 28]
```diff
@@ -268,12 +269,154 @@ func libComposeToKomposeMapping(composeObject *project.Project) (kobject.Kompose
...
+		for _, depSvc := range komposeObject.ServiceConfigs[svcName].VolumesFrom {
+			// recursive call for retrieving volumes of services from `volumes-from`
+			dVols := retrieveVolume(depSvc, komposeObject)
+			var cVols []kobject.Volumes
+			// creating array of volumes of current service(svcName)
+			for _, v := range komposeObject.ServiceConfigs[svcName].VolList {
+				vol := kobject.Volumes{
+					SvcName:   svcName,
+					MountPath: v,
+				}
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: You can use `transformer.ParseVolume` here, then you won't need to call it later in the final volume.

File: pkg/loader/compose/v1v2.go (hunk 26 above)
Comment (surajnarwade) [inline pkg/loader/compose/v1v2.go]: it will put further output in same struct `FinalVol`

File: pkg/loader/compose/v1v2.go (hunk 28 above)
Comment (surajnarwade) [inline pkg/loader/compose/v1v2.go]: but it will work only when `volumes-from` is not present

File: pkg/loader/compose/v1v2.go (hunk 27 above)
Comment (surajnarwade) [inline pkg/loader/compose/v1v2.go]: but it will work only when volumes-from is not present

File: pkg/loader/compose/v1v2.go (hunk 28 above)
Comment (surajnarwade) [inline pkg/loader/compose/v1v2.go]: would it be better to call at later stage than putting this twice here ?

File: pkg/loader/compose/v1v2.go (hunk 26 above)
Comment (surajnarwade) [inline pkg/loader/compose/v1v2.go]: even we are naming PVC here

File: pkg/loader/compose/v1v2.go (hunk 27 above)
Comment (surajnarwade) [inline pkg/loader/compose/v1v2.go]: PVC naming can't be done here

File: pkg/loader/compose/v1v2.go [hunk 29]
```diff
@@ -268,12 +269,154 @@ func libComposeToKomposeMapping(composeObject *project.Project) (kobject.Kompose
...
+			}
+
+		}
+
+		// We can't assign value to struct field in map while iterating over it, so temporary variable `temp` is used here
+
+		var temp = komposeObject.ServiceConfigs[name]
+		temp.Volumes = FinalVol
+		komposeObject.ServiceConfigs[name] = temp
+		for _, service := range komposeObject.ServiceConfigs {
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: why is this for loop in the same kinda for loop? I don't see reason of running this for loop inside https://github.com/kubernetes/kompose/pull/626/files#diff-61f73b3eaf591a7fc27f6c3602535a9aR281

It is running unnecessarily for many iterations.

File: pkg/loader/compose/v1v2.go [hunk 30]
```diff
@@ -268,12 +269,154 @@ func libComposeToKomposeMapping(composeObject *project.Project) (kobject.Kompose
...
+			for v, _ := range service.Volumes {
+				vol := &service.Volumes[v]
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: no need to get the address just use the index everywhere like `service.Volumes[v].VFrom`...

File: pkg/loader/compose/v1v2.go [hunk 31]
```diff
@@ -268,12 +269,123 @@ func libComposeToKomposeMapping(composeObject *project.Project) (kobject.Kompose
 			log.Infof("Service name in docker-compose has been changed from %q to %q", name, normalizeServiceNames(name))
 		}
 	}
+
+	// This will handle volume at earlier stage itself, it will resolves problems occurred due to `volumes_from` key
+	handleVolume(&komposeObject)
```
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: we can get rid of this function by moving all of its code in the for loop above

**Comment by surajssd** [comment]
Also add the test case in functional tests that you are using for testing locally it has some good corner case checks, also maybe name them something better!

File: pkg/loader/compose/v1v2.go (hunk 31 above)
Comment (surajnarwade) [inline pkg/loader/compose/v1v2.go]: @surajssd , loop iterate over a service one at a time. as we are dealing with `volumes_from` key here, we can't put this line into for loop

**Comment by surajnarwade** [comment]
@surajssd , I have updated volume_from example which covers all corner cases probably.

**Comment by surajnarwade** [comment]
@kadel , needs your review :)

File: script/test/fixtures/volume-mounts/volumes-from/docker-compose.yml [hunk 32]
```diff
@@ -1,20 +1,32 @@
...
-      - web
+       - afoo
+
+  cat:
+    image: busybox
+    command: sleep 3600
+    volumes:
+      - /cat
+      - /catsuraj
+
```
Comment (cdrage) [inline script/test/fixtures/volume-mounts/volumes-from/docker-compose.yml]: Please don't change the current test example.

Can you please take the above code and create a different, separate example from this?

File: pkg/loader/compose/v1v2.go (hunk 31 above)
Comment (surajssd) [inline pkg/loader/compose/v1v2.go]: yeah makes sense! I didn't realize that before!

**Comment by surajnarwade** [comment]
@cdrage , added new example

File: script/test/fixtures/volume-mounts/volumes-from/docker-compose.yml [hunk 33]
```diff
@@ -17,4 +17,4 @@ services:
     volumes:
       - /www/public
     volumes_from:
-      - web
+      - web
```
Comment (cdrage) [inline script/test/fixtures/volume-mounts/volumes-from/docker-compose.yml]: added space I think?

Comment (surajnarwade) [inline script/test/fixtures/volume-mounts/volumes-from/docker-compose.yml]: I don't think so

Comment (cdrage) [inline script/test/fixtures/volume-mounts/volumes-from/docker-compose.yml]: you've got to do `git checkout script/test/fixtures/volume-mounts/volumes-from/docker-compose.yml` to revert it back to normal. you've modified something here (newline maybe?)

Comment (surajnarwade) [inline script/test/fixtures/volume-mounts/volumes-from/docker-compose.yml]: @cdrage already resolved that change, so it's outdated now :)

**Comment by cdrage** [comment]
LGTM!

File: pkg/loader/compose/v1v2.go [hunk 34]
```diff
@@ -268,12 +269,123 @@ func libComposeToKomposeMapping(composeObject *project.Project) (kobject.Kompose
...
+		v.SvcName = svcName
+		v.MountPath = vn
+		v.PVCName = fmt.Sprintf("%s-claim%d", v.SvcName, i)
+		volumes = append(volumes, v)
+	}
+	return volumes, nil
+}
+
+// for current volumes
+func checkVolPresent(toFind kobject.Volumes, Vols []kobject.Volumes) bool {
```
Comment (kadel) [inline pkg/loader/compose/v1v2.go]: Shouldn't this be named `checkVolNotPresent`?

**Comment by kadel** [comment]
> It will resolve #544 as well as refactor volume handling part.

this is not resolving #544 is it?
Isn't this meant to be just ground work for resolving #544 in the future?

**Comment by surajnarwade** [comment]
@kadel , it will completely resolve #544

**Comment by kadel** [comment]
> @kadel , it will completely resolve #544

I've tested that and result is still the same :-(

**Comment by surajnarwade** [comment]
@kadel , problem is occuring due to permissions of volumes in the example mentioned in the issue.

**Comment by kadel** [comment]
> @kadel , problem is occuring due to permissions of volumes in the example mentioned in the issue.

yeh, but it should work even with permissions. 
If volumes are the same, only permissions differ that kompose has to set `readOnly` for `ro` volumes. Right now it puts duplicate volume there.

File: pkg/loader/compose/v1v2.go [hunk 35]
```diff
@@ -268,12 +269,123 @@ func libComposeToKomposeMapping(composeObject *project.Project) (kobject.Kompose
...
+	var err error
+
+	for i, vn := range volNames {
+		var v kobject.Volumes
+		v.VolumeName, v.Host, v.Container, v.Mode, err = transformer.ParseVolume(vn)
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not parse volume %q: %v", vn, err)
+		}
+		v.SvcName = svcName
+		v.MountPath = vn
```
Comment (kadel) [inline pkg/loader/compose/v1v2.go]: `vn` contains full path with permissions. like `./certs:/etc/nginx/certs:ro`.
Shouldn't permission be stripped from it before saving it to `MountPath`?

It might solve problem with duplicate volumes with different permissions https://github.com/kubernetes/kompose/pull/626#issuecomment-316718946

Comment (kadel) [inline pkg/loader/compose/v1v2.go]: it can be quickly fixed like this:  `v.MountPath = fmt.Sprintf("%s:%s", v.Host, v.Container)`

This made me realize that we don't even need `v.MountPath` as it can be always assembled from `v.Host` +  `v.Container` and if needed `v.Mode`

**Comment by surajnarwade** [comment]
@kadel , now it works

**Comment by kadel** [comment]
@surajnarwade  getting closer ;-)

But permissions are still not correct:-(

letsencrypt Deployment has `/etc/nginx/certs` mounted as readOnly, even  if docker-compose says `rw`

```yaml
          - mountPath: /etc/nginx/certs
            name: proxy-claim2
            readOnly: true
```

**Comment by cdrage** [comment]
@kadel Odd that this is not caught in a failing test, did @surajnarwade override it or is there actually no tests written for checking if the readOnly value has been enabled?

**Comment by kadel** [comment]
> @kadel Odd that this is not caught in a failing test, did @surajnarwade override it or is there actually no tests written for checking if the readOnly value has been enabled?

It looks like there is no test case checking that :-(

There should be a test testing this case: One volume mounted to two different containers, in one container as ro, in other as rw

File: pkg/loader/compose/v1v2.go [hunk 36]
```diff
@@ -268,12 +269,123 @@ func libComposeToKomposeMapping(composeObject *project.Project) (kobject.Kompose
...
+			return true
+		}
+	}
+	return false
+}
+
+// for dependent volumes, returns true and the respective volume if mountpath are same
+func getVol(toFind kobject.Volumes, Vols []kobject.Volumes) (bool, kobject.Volumes) {
+	for _, dv := range Vols {
+		if toFind.MountPath == dv.MountPath && toFind.Mode == dv.Mode {
```
Comment (kadel) [inline pkg/loader/compose/v1v2.go]: `toFind.Mode == dv.Mode` shouldn't be here. It bring back issue with duplicate volumeMounts

**Comment by kadel** [comment]
I found another issue with this.

if you have docker-compose like this:

```yaml
version: "2"

services:
  foo:
    image: quay.io/tomkral/sleeper
    volumes:
     - /opt/data
    
  bar:
    image: quay.io/tomkral/sleeper
    volumes:
     - /opt/data
  ```

`/opt/data` shouldn't be shared between those containers, every container should have its own PVC for `/opt/data` in this case.

**Comment by kadel** [comment]
I've created simple test file:


```yaml
version: "2"

services:
  foo:
    image: quay.io/tomkral/sleeper
    volumes:
     - ./data:/opt/data:ro
    
  bar:
    image: quay.io/tomkral/sleeper
    volumes_from: 
     - foo
    volumes:
     - ./data:/opt/data:rw
```

using this I've found weird bug
If I use this docker-compose file I Deploymnet for bar service doesn't have any volumeMounts.

Once I add second volume for bar service:
```
    volumes:
     - ./data:/opt/data:rw
     - /tmp
```
generated Deployment suddenly has both volumeMounts

**Comment by kadel** [comment]
I think that for this docker-compose 
```yaml
version: "2"

services:
  foo:
    image: quay.io/tomkral/sleeper
    volumes:
     - ./data:/opt/data:ro
    
  bar:
    image: quay.io/tomkral/sleeper
    volumes_from: 
     - foo
    volumes:
     - ./data:/opt/data:rw
```

correct output should be

```yaml
apiVersion: v1
items:
- apiVersion: v1
  kind: Service
  metadata:
    creationTimestamp: null
    labels:
      io.kompose.service: bar
    name: bar
  spec:
    clusterIP: None
    ports:
    - name: headless
      port: 55555
      targetPort: 0
    selector:
      io.kompose.service: bar
  status:
    loadBalancer: {}
- apiVersion: v1
  kind: Service
  metadata:
    creationTimestamp: null
    labels:
      io.kompose.service: foo
    name: foo
  spec:
    clusterIP: None
    ports:
    - name: headless
      port: 55555
      targetPort: 0
    selector:
      io.kompose.service: foo
  status:
    loadBalancer: {}
- apiVersion: extensions/v1beta1
  kind: Deployment
  metadata:
    creationTimestamp: null
    labels:
      io.kompose.service: bar
    name: bar
  spec:
    replicas: 1
    strategy:
      type: Recreate
    template:
      metadata:
        creationTimestamp: null
        labels:
          io.kompose.service: bar
      spec:
        containers:
        - image: quay.io/tomkral/sleeper
          name: bar
          resources: {}
          volumeMounts:
          - mountPath: /opt/data
            name: foo-claim0
        restartPolicy: Always
        volumes:
        - name: foo-claim0
          persistentVolumeClaim:
            claimName: foo-claim0
  status: {}
- apiVersion: extensions/v1beta1
  kind: Deployment
  metadata:
    creationTimestamp: null
    labels:
      io.kompose.service: foo
    name: foo
  spec:
    replicas: 1
    strategy:
      type: Recreate
    template:
      metadata:
        creationTimestamp: null
        labels:
          io.kompose.service: foo
      spec:
        containers:
        - image: quay.io/tomkral/sleeper
          name: foo
          resources: {}
          volumeMounts:
          - mountPath: /opt/data
            name: foo-claim0
            readOnly: true
        restartPolicy: Always
        volumes:
        - name: foo-claim0
          persistentVolumeClaim:
            claimName: foo-claim0
            readOnly: true
  status: {}
- apiVersion: v1
  kind: PersistentVolumeClaim
  metadata:
    creationTimestamp: null
    labels:
      io.kompose.service: foo-claim0
    name: foo-claim0
  spec:
    accessModes:
    - ReadWriteOnce
    resources:
      requests:
        storage: 100Mi
  status: {}
kind: List
metadata: {}

```

**Comment by kadel** [comment]
It looks like all the problems were solved in the last update of this PR. :tada: 

Grat work @surajnarwade ! :+1:
